runtime/lua/vim/lsp/sync.lua

```diff
@@ -0,0 +1,384 @@
+-- Notes on incremental sync:
+--  Per the protocol, the text range should be:
+--
+--  A position inside a document (see Position definition below) is expressed as
+--  a zero-based line and character offset. The offsets are based on a UTF-16
+--  string representation. So a string of the form a𐐀b the character offset
+--  of the character a is 0, the character offset of 𐐀 is 1 and the character
+--  offset of b is 3 since 𐐀 is represented using two code units in UTF-16.
+--
+--  To ensure that both client and server split the string into the same line
+--  representation the protocol specifies the following end-of-line sequences: ‘\n’, ‘\r\n’ and ‘\r’.
+--
+--  Positions are line end character agnostic. So you can not specify a position that
+--  denotes \r|\n or \n| where | represents the character offset. This means *no* defining
+--  a range than ends on the same line after a terminating character
+--
+-- Generic warnings about byte level changes in neovim. Many apparently "single"
+-- operations in on_lines callbacks are actually multiple operations.
+--
+--  Join operation (2 operations):
+--  * extends line 1 with the contents of line 2
+--  * deletes line 2
+--
+--  test 1    test 1 test 2    test 1 test 2
+--  test 2 -> test 2        -> test 3
+--  test 3    test 3
+--
+--  Deleting (and undoing) two middle lines (1 operation):
+--
+--  test 1    test 1
+--  test 2 -> test 4
+--  test 3
+--  test 4
+--
+--  Deleting partial lines (5 operations) deleting between asterisks below:
+--
+--  test *1   test *    test *     test *    test *4    test *4*
+--  test 2 -> test 2 -> test *4 -> *4     -> *4      ->
+--  test 3    test 3
+--  test *4   test 4
+
+local M = {}
+
+-- local string.byte, unclear if this is necessary for JIT compilation
+local str_byte = string.byte
+local vim = vim
+local math = math
```

Comment:
Does this make a difference?

Reply:
I believe making the tables local avoid a global lookup in each access, I'm not 100% positive on this though.

Reply:
It does, according to Mike (bullet point 7) http://wiki.luajit.org/Numerical-Computing-Performance-Guide

Reply:
Should I cache math.min as math_min instead?

Reply:
Added some more aggresive local caching https://github.com/neovim/neovim/pull/16252/commits/72f27ab529764a2ae65d2eea654fc687b9fb99a4

Reply:
> It does, according to Mike (bullet point 7) http://wiki.luajit.org/Numerical-Computing-Performance-Guide

Afaik that's a bit different than the `local vim = vim` case here - It assigns something within a table to the local to avoid the table lookups. 

> Added some more aggresive local caching 72f27ab

:+1:

Reply:
I think my final update is in line with the performance guide though right? (no table lookup)

Reply:
As far as I understand - yes.